### PR TITLE
add full IIM support to IBFESurfaceMethod (part 4)

### PR DESCRIPTION
**Pull request by boyceg**
### IBAMR Pull Request Checklist
- [ ] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [ ] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

**Comment by boyceg** [comment]
@AminKolah, I created a new example, ex12, for the IIM code.

Note that the example currently crashes -- there are a number of places where the code assumes that the jump condition-related data structures will be allocated. Could you take a look and track down remaining issues? Add a commit with the fixes (please don't force push rebased changes) so that I can backport to parts 1-3.

The code needs to work for all combinations of jump conditions enabled/disabled, and should only allocate jump condition data for a particular type of jump condition when that jump condition is enabled.

**Comment by AminKolah** [comment]
> @AminKolah, I created a new example, ex12, for the IIM code.
> 
> Note that the example currently crashes -- there are a number of places where the code assumes that the jump condition-related data structures will be allocated. Could you take a look and track down remaining issues? Add a commit with the fixes (please don't force push rebased changes) so that I can backport to parts 1-3.
> 
> The code needs to work for all combinations of jump conditions enabled/disabled, and should only allocate jump condition data for a particular type of jump condition when that jump condition is enabled.

Okay I'm looking at it now...

**Comment by AminKolah** [comment]
> @AminKolah, I created a new example, ex12, for the IIM code.
> 
> Note that the example currently crashes -- there are a number of places where the code assumes that the jump condition-related data structures will be allocated. Could you take a look and track down remaining issues? Add a commit with the fixes (please don't force push rebased changes) so that I can backport to parts 1-3.
> 
> The code needs to work for all combinations of jump conditions enabled/disabled, and should only allocate jump condition data for a particular type of jump condition when that jump condition is enabled.


I want to push some commits here but I don't want to mess anything up with the whole force push thing. Can I create a branch from this to push the changes quickly? You can take a look and merge that into the current one.

**Comment by boyceg** [comment]
Please just push to this branch.

**Comment by boyceg** [comment]
@AminKolah you cannot use preprocessor directives in the way that you are trying to use them here in these last two commits. You need to use regular conditionals.

**Comment by drwells** [comment]
Is the intent of this branch to having a working implementation of IIM and just merge small features on 1-3?

**Comment by boyceg** [comment]
The four branches together include all of the specific IIM features from @AminKolah's branch. There is dependency between them --- they have to be merged in order.

File: src/IB/IBFESurfaceMethod.cpp [hunk 1]
```diff
@@ -430,28 +430,44 @@ IBFESurfaceMethod::preprocessIntegrateData(double current_time, double new_time,
...
+                //~ d_WSS_in->half_vecs, d_WSS_out->half_vecs, d_P_in->half_vecs, d_P_out->half_vecs, d_TAU_in->half_vecs,
+                //~ d_TAU_out->half_vecs, d_DU_jump[0]->half_vecs, d_DU_jump[1]->half_vecs,
+//~ #if (NDIM == 3)
+                //~ d_DU_jump[2]->half_vecs
+//~ #endif
+        //~ },
+        //~ INSERT_VALUES,
+        //~ SCATTER_FORWARD);
+        
+	std::vector<std::vector<libMesh::PetscVector<double>*> > Vec_collection_update = {d_U->new_vecs,d_X->new_vecs,d_U_n->new_vecs,
```
Comment (boyceg) [inline src/IB/IBFESurfaceMethod.cpp]:
```suggestion
	std::vector<std::vector<libMesh::PetscVector<double>*> > vec_collection_update = {d_U->new_vecs,d_X->new_vecs,d_U_n->new_vecs,
```

File: src/IB/IBFESurfaceMethod.cpp [hunk 2]
```diff
@@ -430,28 +430,44 @@ IBFESurfaceMethod::preprocessIntegrateData(double current_time, double new_time,
...
-        {
-            d_X->new_vecs, d_U->new_vecs, d_U_n->new_vecs, d_U_t->new_vecs, d_F->half_vecs
-#if (d_use_pressure_jump_conditions)
-                ,
-                d_P_jump->half_vecs, d_P_in->half_vecs, d_P_out->half_vecs
-#endif
-#if (d_use_velocity_jump_conditions)
-                ,
-                d_WSS_in->half_vecs, d_WSS_out->half_vecs, d_DU_jump[0]->half_vecs, d_DU_jump[1]->half_vecs
+    //~ batch_vec_ghost_update(
```
Comment (boyceg) [inline src/IB/IBFESurfaceMethod.cpp]: Just get rid of this --- don't comment it out.

File: src/IB/IBFESurfaceMethod.cpp [hunk 3]
```diff
@@ -430,28 +430,44 @@ IBFESurfaceMethod::preprocessIntegrateData(double current_time, double new_time,
...
+                //~ d_DU_jump[2]->half_vecs
+//~ #endif
+        //~ },
+        //~ INSERT_VALUES,
+        //~ SCATTER_FORWARD);
+        
+	std::vector<std::vector<libMesh::PetscVector<double>*> > Vec_collection_update = {d_U->new_vecs,d_X->new_vecs,d_U_n->new_vecs,
+																						d_U_t->new_vecs, d_F->half_vecs};
+																						
+	if (d_use_pressure_jump_conditions)
```
Comment (boyceg) [inline src/IB/IBFESurfaceMethod.cpp]: 👍

File: src/IB/IBFESurfaceMethod.cpp [hunk 4]
```diff
@@ -430,28 +430,44 @@ IBFESurfaceMethod::preprocessIntegrateData(double current_time, double new_time,
...
-        },
-        INSERT_VALUES,
-        SCATTER_FORWARD);
+	}
+	if (d_compute_fluid_traction)
+	{
+		Vec_collection_update.push_back(d_TAU_in->half_vecs);
+		Vec_collection_update.push_back(d_TAU_out->half_vecs);	
+	}			
+	batch_vec_ghost_update({Vec_collection_update}, INSERT_VALUES, SCATTER_FORWARD);
```
Comment (boyceg) [inline src/IB/IBFESurfaceMethod.cpp]:
```suggestion
	batch_vec_ghost_update(Vec_collection_update, INSERT_VALUES, SCATTER_FORWARD);
```

File: src/IB/IBFESurfaceMethod.cpp (hunk 2 above)
Comment (AminKolah) [reply src/IB/IBFESurfaceMethod.cpp]: aah i wiped it out but apparently I committed before saving!

**Comment by AminKolah** [comment]
@boyceg the issue with different jump options and the solution is fixed now. Would you take a look and let me know whenever you'd like to merge the branch.

**Pull request closed without merging**